proposals/code-signing-for-homebrew.md

```diff
@@ -0,0 +1,101 @@
+
+[Homebrew](https://brew.sh/) is the predominant package manager for macOS, with millions of daily users and [hundreds of active contributors](https://github.com/Homebrew/brew/graphs/contributors). Homebrew is written in Ruby, and uses GitHub Actions to create binary packages (called "[bottles](https://docs.brew.sh/Bottles)") of popular open source projects, which clients then install on their machines via the `brew` CLI. Homebrew successfully delivers [over 400 million](https://formulae.brew.sh/analytics/install/365d/) binary builds of open source packages to macOS users each year.
+
+[Sigstore](https://www.sigstore.dev/) is a Linux Foundation project aimed at making codesigning available to everybody. It uses a combination of open standards (OpenID Connect, WebPKI, Certificate Transparency) to bind non-cryptographic identities (such as emails or GitHub Actions workflow identities) to short-lived signing keys, which can then be used to sign for artifacts much like `gpg` or any other signing tool.
+
+## Technical Items
+
+### Sigstore signatures for official Homebrew taps
+
+Homebrew's packaging processes are built around GitHub Actions. In particular, Homebrew uses GitHub Actions to perform builds of all core formulae (package specifications, like LLVM, `bash`, or `git`) into bottles (binary distributions that can be installed on client machines, similar to an `.app` or `.pkg`). Bottles are then hosted on GitHub Packages (or other hosts, for third-party taps) and are installed by end-user clients e.g. during `brew install` invocations.
+
+Because Homebrew uses GitHub Actions for all `homebrew-core` builds, the Homebrew CI has trivial access to GitHub Actions' [ambient OpenID Connect identities](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect). These identities can in turn be used to sign with Sigstore, meaning that Homebrew's existing CI workflows would require relatively little additional configuration (and zero external key management) to produce valid Sigstore signatures for each bottle build.
+
+The high-level flow for bottle signing:
+
+1. Existing bottle-building workflows will gain a signing step, which will only run as a finalization step for bottles that appear on `master` (i.e., not just the temporary bottles that are built as part of the normal PR lifecycle).
+2. The signing step is responsible for producing a Sigstore signature that _attests_ to the bottle. This should include, at minimum, the bottle's SHA256 digest and the bottle's filename. The digest is necessary to produce a proof that the signature is for a particular bottle's contents, and the filename is necessary to produce a proof that the signature is intended to accompany a particular install step (i.e. `brew install foo` should not verify if an attacker substitutes `foo`'s bottle for a correctly signed bottle of bar, or even a different version of foo than the resolver produces).
```

Comment:
I think @di will need to update the PR here, but to clarify: the thing being signed over is an attestation over (at least) the following:

1. A digest of the bytes for the bottle archive, i.e. `H(read("some-bottle-xxx.tar.gz"))`
2. The bottle's base filename (to prevent domain confusion between signatures/versions of bottles)

My thought was to define that as a custom payload, since Homebrew's bottles aren't *actually* OCI manifests -- they just happen to be stored that way in `homebrew-core`'s storage, since they reuse GitHub Packages.

Reply:
(@woodruffw I can update but won't be able to get to this today, if you want to make a PR against this branch I'll be happy to merge changes that way)

Reply:
https://github.com/ossf/wg-securing-software-repos/pull/21/commits/fc958da7a7ce99330ec7a7ed711d0379b89ccb89

Reply:
Could we not use cosign style blob signing like they do for wasm and other blobby things?

Reply:
> Could we not use cosign style blob signing like they do for wasm and other blobby things?

That's the plan! The "blob" in this case is the attestation payload; my apologies if "payload" was misleading.

(More specifically, my thought here was to format the attestation as an `in-toto` attestation document, reusing/adapting the existing NPM format if possible, and then doing "blob" signing over that. That might need some slight tweaking e.g. to ensure that we submit `intoto` rather than `hashedrekord` to Rekor, but that's the broad idea.)

Reply:
I think @lukehinds was asking "why not just sign over the blob's bytes" vs. signing over some attestation payload, and the question at the root is probably: "why does the filename need to be included in the attestation?"

Reply:
Oh, gotcha. The reason for signing over the filename as well is to prevent (unintended) domain separation: two potential attack scenarios here:

1. A "mix and match" where an attacker substitutes one correctly signed package version for another (older, vulnerable) one. Without the filename (which includes the version) our signature check would accept that _any_ previously signed package or package version.
2. Cache poisoning or confusion attacks where an attacker manages to confuse Homebrew into signing for and/or serving unintended bottles; signatures over these bottles would be rejected so long as the attestation includes the filename, since they wouldn't match the user's expected bottle.

Reply:
Have you considered using [a VSA](https://slsa.dev/verification_summary/v1) as the attestation type?  The filename to be checked would go in `resourceUri`, you could use `verifiedLevels` to distinguish between things that have been newly built vs those that have been backfilled.  Note that VSA is likely [moving out of the SLSA repo](https://github.com/slsa-framework/slsa/issues/921#issuecomment-1638535179) and into the in-toto repo to be made more generic.

Reply:
I might be misunderstanding, but is VSA meant for SLSA provenance _generators_ or for _verifiers_ of that provenance? From that link, it looks like the latter (since it's a countersignature/counterattestation by another party), whereas this proposal is focusing on the former (generating the initial attestation of provenance).

If the above is correct, then I _think_ we use a non-VSA predicate type for the attestations that Homebrew itself is producing, but may want to use the VSA type for the "source attestation" idea mentioned later in the proposal (where Homebrew is verifying a downstream project's source artifact before producing its own build provenance).

(Apologies if I'm way off base in my understanding here!)

Reply:
IMO the VSA can be used anytime you want a downstream user to be able to trust some upstream system's opinion on the 'goodness' of an artifact.  That upstream system could be the same as the one generating the provenance or it could be different.  Either way the VSA is a convenient abstraction for downstream users that don't actually want to evaluate the entire provenance (similar to what code-signing does but with slightly more metadata).

For safety reasons I'd typically expect the system creating the SLSA provenance to be separate from the one doing code-signing or VSA production, but it doesn't have to be.

Using a VSA could also be a straightforward mechanism to resolve point 1 from [your comment here](https://github.com/ossf/wg-securing-software-repos/pull/20#issuecomment-1640397166).  VSAs could be issued for all those old things, either making _no_ statement in `verifiedLevels` or explicitly adding something like `SLSA_BUILD_LEVEL_UNEVALUATED` to make it clear it's not saying anything about how it was built, but still providing the basic protections of code signing.  This is somewhat similar to [the VSA use case described here](https://security.googleblog.com/2022/04/how-to-slsa-part-1-basics.html#:~:text=of%20different%20times.-,On%20import%20to%20repo,-Squirrel%20plans%20to) and [here](https://security.googleblog.com/2022/04/how-to-slsa-part-3-putting-it-all.html#:~:text=Downstream%20users%20of%20this%20Acme%20container%20can%20then%20check%20the%20Acme%20issued%20VSA).